In [5]:
import pandas as pd

pd.set_option('display.max_colwidth', None) #Print full text
pd.set_option('display.max_rows', 1000) #Print full text

### Compute overall statistic on newspapers and speakers
Important no threshold has been applied on speaker

In [25]:
df = pd.read_csv(NEWSPAPER_SPEAKER_COUNT_FILE,compression='bz2')
df.columns

Index(['newspaper', 'speaker', 'count'], dtype='object')

In [9]:
print(f'They are {len(df)} pairs of (newspapers,speakers)')
newspapers = set(df['newspaper'])
print(f'They are {len(newspapers)} unique newspapers')
speakers = set(df['speaker'])
print(f'They are {len(speakers)} unique speakers')

They are 2434410 pairs of (newspapers,speakers)
They are 7362 unique newspapers
They are 212146 unique speakers


In [10]:
speaker_unique_occurence = df[['newspaper','speaker']].groupby('speaker',as_index=False).count()\
                    .rename(columns = {'newspaper':'unique_newspaper_quotes'})

In [11]:
speaker_unique_occurence.nlargest(100,columns=['unique_newspaper_quotes'])

speaker  unique_newspaper_quotes
149814                        none                     6999
161395      president donald trump                     2836
161405             president trump                     2034
194583  tedros adhanom ghebreyesus                     1954
24574                boris johnson                     1890
20952               bernie sanders                     1816
14547                anthony fauci                     1755
10674                 andrew cuomo                     1707
94541                    joe biden                     1683
145222                nancy pelosi                     1665
54377                 donald trump                     1602
58674             elizabeth warren                     1557
141214                  mike pence                     1477
141246                 mike pompeo                     1467
37588                chuck schumer                     1445
114032                 kobe bryant                     1432
67637                 gavin newsom                     1381
142390             mitch mcconnell                     1362
157133              pete buttigieg                     1351
190349              steven mnuchin                     1348
106301              justin trudeau                     1332
21901               bill de blasio                     1297
1902                   adam schiff                     1255
180703              scott morrison                     1248
207768                william barr                     1181
128770                  mark esper                     1176
184304                  shinzo abe                     1176
53615                 dominic raab                     1168
5212                     alex azar                     1167
209345                  xi jinping                     1166
136364           michael bloomberg                     1161
9082                 amy klobuchar                     1154
132876                matt hancock                     1145
140244              mike bloomberg                     1128
90908                 jerome adams                     1104
161028                pope francis                     1102
6856                  ali khamenei                     1101
187639           stephanie grisham                     1100
142844        mohammad javad zarif                     1096
142509                 mitt romney                     1086
75196               hassan rouhani                     1067
199324                   tom hanks                     1062
120363              lindsey graham                     1055
15180             antonio guterres                     1054
59819              emmanuel macron                     1053
50258                 deborah birx                     1042
205595              vladimir putin                     1016
90940                jerome powell                      979
195721                 thomas bach                      977
135445               melania trump                      970
117671                lebron james                      968
141365                   mike ryan                      964
145474               narendra modi                      960
70508               giuseppe conte                      957
12591                angela merkel                      948
88529                   jeff bezos                      942
20461           benjamin netanyahu                      925
203237        ursula von der leyen                      923
161388      president barack obama                      917
3976               alan dershowitz                      911
72835               greta thunberg                      907
161802          queen elizabeth ii                      900
192072               susan collins                      899
161456                prince harry                      898
94317              joaquin phoenix                      894
115712             lamar alexander                      892
87522  

In [12]:
speaker_unique_occurence.describe()

unique_newspaper_quotes
count            212146.000000
mean                 11.475163
std                  44.151353
min                   1.000000
25%                   1.000000
50%                   2.000000
75%                   6.000000
max                6999.000000

In [13]:
speaker_total_occurence = df.groupby('speaker',as_index=False).aggregate({'count':'sum'})\
                                .rename(columns = {'count':'total_newspaper_quotes'})

In [24]:
speaker_total_occurence.nlargest(16000,columns=['total_newspaper_quotes'])

speaker  total_newspaper_quotes
149814                    none                 5857313
161395  president donald trump                  261722
24574            boris johnson                   86315
94541                joe biden                   84953
20952           bernie sanders                   69722
...                        ...                     ...
69736             gerry murphy                      97
70782             glenn martin                      97
70938              glyn hodges                      97
71484             gracie elvin                      97
71653          graham coughlan                      97

[16000 rows x 2 columns]

In [15]:
speaker_total_occurence.describe()

total_newspaper_quotes
count            2.121460e+05
mean             8.040466e+01
std              1.274325e+04
min              1.000000e+00
25%              2.000000e+00
50%              6.000000e+00
75%              1.900000e+01
max              5.857313e+06

In [16]:
newspaper_speaker_count = df.groupby('newspaper',as_index=False).aggregate({'speaker':'count'})\
                            .rename(columns = {'speaker':'speaker_count'})

In [17]:
newspaper_speaker_count.nlargest(100,columns=['speaker_count'])

newspaper  speaker_count
4055                   msn          16544
784              breitbart           9434
6872       washingtontimes           9239
5439                sfgate           8864
2954            indiatimes           7304
4518               nytimes           7277
5405          seattletimes           6951
2938           independent           6501
5686      stamfordadvocate           6426
890        businessinsider           6422
1478                ctpost           6411
6140               thehour           6277
4340             newstimes           6222
4378            nhregister           6066
5769                 stuff           6007
2980            infosurhoy           5560
4523                nzcity           5479
3417                  kvia           5468
6041            thecourier           5426
4127          mysanantonio           5358
1178                 chron           5287
5551                   smh           5256
2882               ibtimes           5246
2302               foxnews           5237
803          brisbanetimes           5231
6323               thestar           5182
6118           theguardian           5166
4325                newsok           5142
528                    bbc           5090
5701           startribune           5061
2982               infotel           4938
5966                theage           4897
3918                mirror           4874
1111       channelnewsasia           4826
2481                    go           4713
2060           expressnews           4704
7242                  wtop           4609
744            bostonglobe           4552
7239                  wtmj           4490
2058               express           4487
2059        expressandstar           4487
7165             wordpress           4483
4283                news18           4458
1276                   cnn           4428
2948         indianexpress           4419
3374                   ksl           4368
621         bignewsnetwork           4304
34                 570news           4300
6331                thesun           4157
2593         greenwichtime           4082
3237               kdhnews           4066
1072              cbslocal           4058
5799    sunshinecoastdaily           4055
5689              standard           4046
1480               ctvnews           4043
3075         irishexaminer           4000
850         bucksfreepress           3993
2033          eveningtimes           3993
5921             telegraph           3990
6022          thechronicle           3984
4282                news12           3950
6882               watoday           3903
1211              citynews           3868
779           breakingnews           3848
5028                    qt           3840
4525              nzherald           3829
4280             news-mail           3826
2750        hindustantimes           3821
2626           gympietimes           3816
6209    themorningbulletin           3815
1554          dailymercury           3804
2247                forbes           3776
7338     yourlocalguardian           3775
2308  frasercoastchronicle           3771
564       belfasttelegraph           3748
6132              thehindu           3743
4646            oxfordmail           3740
6864      warwickdailynews           3739
6995                  wftv           3738
1533         dailyexaminer           3737
6113       theglobeandmail           3725
4444          northernstar           3716
7032       whitsundaytimes           3700
6235       thenorthernecho           3675
2868            hypervocal           3674
2451     gladstoneobserver           3662
5608       southwalesargus           3662
6406       thisiswiltshire           3654
5474        shropshirestar           3638
7325             yorkpress           3614
4964             prokerala           3608
5999         theboltonnews           3593
753        bournemouthecho           3591
4122           mynorthwest           3586
3368                  ksat 

In [18]:
newspaper_speaker_count.describe()

speaker_count
count    7362.000000
mean      330.672372
std       799.103581
min         1.000000
25%        11.000000
50%        53.000000
75%       218.000000
max     16544.000000

In [19]:
newspaper_quote_count = df.groupby('newspaper',as_index=False).aggregate({'count':'sum'})\
                        .rename(columns={'count':'quote_count'})

In [20]:
newspaper_quote_count.describe()

quote_count
count    7362.000000
mean     2316.969302
std      7791.179536
min         1.000000
25%        34.000000
50%       221.500000
75%      1117.500000
max    267567.000000

### Thresholding decisions

If we keep everything, the newspaper-speaker_count matrix will be of size:  
newspapers_count x speaker_count = 7148 x 218 413 = 1 561 216 124 which is huge even though the obtained matrix will be sparse.  
We must decide which newspaper to keep  
We must decide which speakers to keep  
We decide probability threshold for a speaker to be correctly attributed to a quote.  